cpp/src/arrow/util/async_generator.h

```diff
@@ -177,6 +179,94 @@ class TransformingGenerator {
   std::shared_ptr<TransformingGeneratorState> state_;
 };
 
+template <typename T>
+class SerialReadaheadGenerator {
+ public:
+  SerialReadaheadGenerator(AsyncGenerator<T> source_generator, int max_readahead)
+      : state_(std::make_shared<State>(std::move(source_generator), max_readahead)) {}
+
+  Future<T> operator()() {
+    if (state_->first) {
+      // Lazy generator, need to wait for the first ask to prime the pump
+      state_->first = false;
+      auto next = state_->source();
+      return next.Then(Callback{state_});
+    }
+
+    // This generator is not async-reentrant.  We won't be called until the last
+    // future finished so we know there is something in the queue
+    auto finished = state_->finished.load();
+    if (finished && state_->readahead_queue.IsEmpty()) {
+      return Future<T>::MakeFinished(IterationTraits<T>::End());
+    }
+
+    auto next_ptr = state_->readahead_queue.FrontPtr();
+    auto next = std::move(**next_ptr);
+    state_->readahead_queue.PopFront();
+
+    auto last_available = state_->spaces_available.fetch_add(1);
+    if (last_available == 0 && !finished) {
+      // Reader idled out, we need to restart it
+      ARROW_RETURN_NOT_OK(state_->Pump(state_));
+    }
+    return next;
+  }
+
+ private:
+  struct State {
+    State(AsyncGenerator<T> source_, int max_readahead)
+        : first(true),
+          source(std::move(source_)),
+          finished(false),
+          spaces_available(max_readahead),
+          readahead_queue(max_readahead) {}
+
+    Status Pump(const std::shared_ptr<State>& self) {
+      // Can't do readahead_queue.write(source().Then(Callback{self})) because then the
+      // callback might run immediately and add itself to the queue before this gets added
+      // to the queue messing up the order
+      auto next_slot = std::make_shared<Future<T>>();
+      auto written = readahead_queue.Write(next_slot);
+      if (!written) {
+        return Status::UnknownError("Could not write to readahead_queue");
+      }
+      *next_slot = source().Then(Callback{self});
+      return Status::OK();
+    }
+
+    // Only accessed by the consumer end
+    bool first;
+    // Accessed by both threads
+    AsyncGenerator<T> source;
+    std::atomic<bool> finished;
+    std::atomic<uint32_t> spaces_available;
+    util::SpscQueue<std::shared_ptr<Future<T>>> readahead_queue;
```

Comment:
`Future` is already basically a `shared_ptr`. Why do we need to wrap it in another one?

Reply:
We have to set the value of the future after it is added to the queue.  We want the value to be the value returned by `source().Then()` but we can't run `Then()` until after we add it to the queue.  Otherwise, if `Then` triggers synchronously it adds itself to the queue and we get items out of order.

So we use a queue of shared pointers so we can add a pointer to the queue and set the value of that pointer later.  That's also why it's called `next_slot` instead of `next` because we are adding a slot for the future and then we go and get the future.

Reply:
The micro-optimizer in me would consider using `optional<Future<T>>` instead :-)

Reply:
The write would copy/move the optional so I wouldn't be modifying the same bits.  I could add a `BackPtr` method to the queue which would allow this.  Technically then I would be accessing bits that the queue might be using but it shouldn't be accessing the value and it's a non-resizable buffer so it isn't like the queue moves it around or touches it otherwise so I think it would be safe.  Does that make sense?  If so I can address it later today.

Reply:
No, we can keep it like this instead of completing things. But this also seems to be an argument in favour of our own queue class with an adequate API.

Reply:
Follow-up added in ARROW-11943